File: languages/py/2023/01_Trebuchet.py
from lib import *

inp = get_input()

WORD_NUMS = "one_two_three_four_five_six_seven_eight_nine".split("_")

def solve(p2=False):
    su=[]
    for l in inp:
        nums = []

        for i in range(len(l)):
            if l[i].isdigit():
                nums.append(l[i])
            if p2:
                for n, w in enumerate(WORD_NUMS):
                    if l.startswith(w, i):
                        nums.append(str(n + 1))
                        break
        if len(nums) == 0:
            print("-")
            return # don't error if run with second example input

        su.append(int(nums[0] + nums[-1]))

    print(sum(su))

solve()
solve(True)

File: languages/js/2023/01_Trebuchet.js
import { textToArray, readInput } from '../lib/utils.js';

const list = textToArray(readInput());

const WORD_NUMS = "one_two_three_four_five_six_seven_eight_nine".split("_");

const solve = (inp, p2 = false) => {
  return inp.map(x => {
    const nums = [];

    for (let i = 0; i < x.length; i++) {
      if (x[i].match(/^\d$/)) {
        nums.push(x[i]);
      }

      if (p2) {
        for (let n = 0; n < WORD_NUMS.length; n++) {
          if (x.startsWith(WORD_NUMS[n], i)) {
            nums.push(`${n + 1}`);
            break;
          }
        }
      }
    }

    return parseInt(nums.at(0) + nums.at(-1));
  }).reduce((a, b) => a + b);
}

console.log(solve(list) || "-") // don't error if run with second example input
console.log(solve(list, true))
